backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt

```diff
@@ -0,0 +1,161 @@
+            val kind = if (library.isInterop) "Interop" else "Default"
+            configuration.report(CompilerMessageSeverity.STRONG_WARNING, illegalKindMessage(kind, library.libraryName))
+        }
+
+        override fun reportNotIncludedLibraries(includedLibraries: List<KonanLibrary>, remainingFeaturedLibraries: Set<File>) {
+            val message = buildString {
+                appendln(notIncludedLibraryMessageTitle())
+                remainingFeaturedLibraries.forEach { appendln(it) }
+                appendln()
+                appendln("Included libraries:")
+                includedLibraries.forEach { appendln(it.libraryFile) }
+            }
+
+            configuration.report(CompilerMessageSeverity.STRONG_WARNING, message)
+        }
+    }
+
+    private class ExportedLibrariesReporter(configuration: CompilerConfiguration) : BaseReporter(configuration) {
+        override fun illegalKindMessage(kind: String, libraryName: String): String =
+            "$kind library $libraryName can't be exported with -Xexport-library"
+
+        override fun notIncludedLibraryMessageTitle(): String =
+            "Following libraries are specified to be exported with -Xexport-library, but not included to the build:"
+    }
+
+    private class SourceLibrariesReporter(configuration: CompilerConfiguration) : BaseReporter(configuration) {
+        override fun illegalKindMessage(kind: String, libraryName: String): String =
+            "$kind library $libraryName can't be used as a source library"
+
+        override fun notIncludedLibraryMessageTitle(): String =
+            "Following libraries are declared as source libraries with -Xsource-library, but not included to the build:"
+    }
+
+    private class CoveredLibraryReporter(configuration: CompilerConfiguration): BaseReporter(configuration) {
+        override fun illegalKindMessage(kind: String, libraryName: String): String =
+            "$kind library $libraryName can't be covered"
+
+        override fun notIncludedLibraryMessageTitle(): String =
+            "Following libraries are specified to be covered with -Xlibrary-to-cover, but not included to the build:"
+    }
+
+    companion object {
+        fun forExportedLibraries(configuration: CompilerConfiguration): FeaturedLibrariesReporter = ExportedLibrariesReporter(configuration)
+        fun forSourceLibraries(configuration: CompilerConfiguration): FeaturedLibrariesReporter = SourceLibrariesReporter(configuration)
+        fun forCoveredLibraries(configuration: CompilerConfiguration): FeaturedLibrariesReporter = CoveredLibraryReporter(configuration)
+    }
+}
+
+private fun getFeaturedLibraries(
+    featuredLibraries: List<String>,
+    resolvedLibraries: KonanLibraryResolveResult,
+    resolver: SearchPathResolver,
+    reporter: FeaturedLibrariesReporter
+) : List<KonanLibrary> {
+    val featuredLibraryFiles = featuredLibraries.toUnresolvedLibraries.map { resolver.resolve(it).libraryFile }.toSet()
```

Comment:
If I understand correctly, resolution here changes the handling (and therefore the meaning) of `-Xexported-library`. Is there a motivation behind this?

Reply:
> resolution here changes the handling (and therefore the meaning) of -Xexported-library

What change do you mean? The only intentional change here is to make `-Xexport-library` more flexible regarding input. In current master only a path to a library file is allowed. In this patch any library description (path or name with optional version) is allowed.

Reply:
Yes, I mean exactly this change.

Reply:
So the only motivation here is to make this flag more flexible. Do you see some issues with it?